### PR TITLE
Add polygon labels

### PR DESCRIPTION
**Pull request by ernstdevreede**
This PR adds support for showing labels with GeoJSON Polygons.
There are some configurable things, like text angle, which property to use as label, label print format, font, font size etc which have defaults at the moment.
I'd like to discuss how and where to add these to the ADAGUC configuration.
For a preview see also: https://www.loom.com/share/ddcd7cbfad3c4f7fb7f239f66a480401

### CODE REVIEW

**Comment by maartenplieger** [comment]
The video looks good! Can you add a functional test to make sure that this feature is tracked with future updates?

File: data/config/includes/DefaultStyles.include.xml [hunk 1]
```diff
@@ -143,16 +143,30 @@
...
     <StandardNames standard_name="*"/>
   </Style>
-  
+
   <Style name="autofeatures">
     <Legend fixedclasses="true" tickround="1">elevation</Legend>
     <Point pointstyle="point" textcolor="#111111"/>
     <Vector vectorstyle="barb" linecolor="#0000FF" />
-    <FeatureInterval match=".*" bgcolor="#CCCCFF" fillcolor="#000080FF"  bordercolor="#000000" borderwidth="2"/> <!-- Used in CImgRenderPolylines -->
+    <FeatureInterval match=".*"
```
Comment (maartenplieger) [inline data/config/includes/DefaultStyles.include.xml:151]: This is for autowms?

File: python/lib/adaguc/runAdaguc.py [hunk 2]
```diff
@@ -14,199 +14,198 @@
...
-
-    def removeLogFile(self):
-        try:
-            os.remove(self.ADAGUC_LOGFILE)
-        except:
-            pass
-
-    def getLogFile(self):
-        try:
-            f = open(self.ADAGUC_LOGFILE,  encoding="utf8", errors='ignore')
```
Comment (maartenplieger) [inline python/lib/adaguc/runAdaguc.py]: Why is this removed?

File: adagucserverEC/CServerConfig_CPPXSD.h [hunk 3]
```diff
@@ -715,6 +717,27 @@ class CServerConfig : public CXMLSerializerInterface {
       } else if (equals("bordercolor", 11, attrname)) {
         attr.bordercolor.copy(attrvalue);
         return;
+      } else if (equals("labelfontfile", 13, attrname)) {
+        attr.labelfontfile.copy(attrvalue);
+        return;
+      } else if (equals("labelfontsize", 13, attrname)) {
```
Comment (maartenplieger) [inline adagucserverEC/CServerConfig_CPPXSD.h:717]: We need to add this to the documentation in the update-docs branch

File: tests/AdagucTests/TestGeoJSONFeatures.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    # Test GetCapabilities
+    filename = "test_GeoJSON_time_GetCapabilities.xml"
+    status, data, headers = AdagucTestTools().runADAGUCServer(
+        "&SERVICE=WMS&&SERVICE=WMS&VERSION=1.3.0&REQUEST=GetCapabilities", env={'ADAGUC_CONFIG': config})
+    AdagucTestTools().writetofile(self.testresultspath + filename, data.getvalue())
+    self.assertEqual(status, 0)
+    self.assertTrue(AdagucTestTools().compareGetCapabilitiesXML(
+        self.testresultspath + filename, self.expectedoutputsspath + filename))
+
+  def test_GeoJSON_3timesteps(self):
```
Comment (maartenplieger) [inline tests/AdagucTests/TestGeoJSONFeatures.py:54]: We need to write a tuturial in the update-docs branch demonstrating this feature.

File: data/config/includes/DefaultStyles.include.xml (hunk 1 above)
Comment (ernstdevreede) [reply data/config/includes/DefaultStyles.include.xml:151]: Good question. It is for autowms, but what should be shown iby autowms. You don't know in advance what to show here.

Any suggestion?

File: python/lib/adaguc/runAdaguc.py (hunk 2 above)
Comment (ernstdevreede) [reply python/lib/adaguc/runAdaguc.py]: No idea. I'll revert it.

File: adagucserverEC/CServerConfig_CPPXSD.h [hunk 5]
```diff
@@ -715,6 +717,27 @@ class CServerConfig : public CXMLSerializerInterface {
...
+        attr.labelfontsize.copy(attrvalue);
```
Comment (maartenplieger) [inline adagucserverEC/CServerConfig_CPPXSD.h:718]: Can you make sure that these are described in https://github.com/KNMI/adaguc-server/blob/add_polygon_labels/doc/configuration/FeatureInterval.md ?

File: data/config/includes/DefaultStyles.include.xml (hunk 1 above)
Comment (maartenplieger) [reply data/config/includes/DefaultStyles.include.xml:151]: I guess it is Ok to do this, but when should autofeatures be added to the stylelist?

File: Dockerfile [hunk 6]
```diff
@@ -83,7 +83,7 @@ RUN pip3 install --no-cache-dir --upgrade pip \
 
 # Run adaguc-server functional and regression tests
 
-RUN bash runtests.sh
+#RUN bash runtests.sh
```
Comment (maartenplieger) [inline Dockerfile]: This one should be enabled again.

File: adagucserverEC/CDrawImage.h [hunk 7]
```diff
@@ -101,6 +101,37 @@ class CColor {
   }
 };
 
+class RectangleText {
```
Comment (maartenplieger) [inline adagucserverEC/CDrawImage.h]: Can you move this one to its own file?

File: adagucserverEC/CGeoJSONData.h [hunk 8]
```diff
@@ -108,7 +108,21 @@ class FeatureProperty {
       s.print("NONE");
     }
     return s;
-  }
+  };
+
+  CT::string toString(const char *fmt) {
```
Comment (maartenplieger) [inline adagucserverEC/CGeoJSONData.h]: Can you keep the declaration here and move the implementation to the CPP file?

File: python/python-adaguc-server/routeAdagucServer.py [hunk 9]
```diff
@@ -13,7 +13,7 @@
 logger = logging.getLogger(__name__)
 
 
-@routeAdagucServer.route("/wms", methods=["GET"])
+@routeAdagucServer.route("/wms", methods=["GET"], strict_slashes=False)
```
Comment (maartenplieger) [inline python/python-adaguc-server/routeAdagucServer.py]: Why did you have to change this?

File: tests/functional_test.py [hunk 10]
```diff
@@ -17,21 +17,23 @@
...
-suites.append(TestLoader().loadTestsFromTestCase(TestWMSDocumentCache))
-suites.append(TestLoader().loadTestsFromTestCase(TestOpenDAPServer))
-suites.append(TestLoader().loadTestsFromTestCase(TestWMSPolylineRenderer))
-suites.append(TestLoader().loadTestsFromTestCase(TestWMSTiling))
-suites.append(TestLoader().loadTestsFromTestCase(TestADAGUCFeatureFunctions))
-suites.append(TestLoader().loadTestsFromTestCase(TestCSV))
-suites.append(TestLoader().loadTestsFromTestCase(TestGeoJSON))
-suites.append(TestLoader().loadTestsFromTestCase(TestMetadataService))
-suites.append(TestLoader().loadTestsFromTestCase(TestWMSVolScan))
+# suites.append(TestLoader().loadTestsFromTestCase(TestWMS))
```
Comment (maartenplieger) [inline tests/functional_test.py]: Can you re-enable these tests again

File: adagucserverEC/CServerConfig_CPPXSD.h (hunk 5 above)
Comment (ernstdevreede) [reply adagucserverEC/CServerConfig_CPPXSD.h:718]: Done

File: Dockerfile (hunk 6 above)
Comment (ernstdevreede) [reply Dockerfile]: Oops.

File: data/config/includes/DefaultStyles.include.xml (hunk 1 above)
Comment (ernstdevreede) [reply data/config/includes/DefaultStyles.include.xml:151]: Not sure what to do with this, as there is no uiniversal default for matchid

File: adagucserverEC/CDrawImage.h (hunk 7 above)
Comment (ernstdevreede) [reply adagucserverEC/CDrawImage.h]: Also had to move CColor to it's own file for this.

File: adagucserverEC/CGeoJSONData.h (hunk 8 above)
Comment (ernstdevreede) [reply adagucserverEC/CGeoJSONData.h]: Yep

File: python/python-adaguc-server/routeAdagucServer.py (hunk 9 above)
Comment (ernstdevreede) [reply python/python-adaguc-server/routeAdagucServer.py]: To make Flask accept trailing slashes (or double slashes).
Not very relevant if we go to FastApi

File: tests/functional_test.py (hunk 10 above)
Comment (ernstdevreede) [reply tests/functional_test.py]: Double oops

File: python/python-adaguc-server/routeAdagucServer.py (hunk 9 above)
Comment (ernstdevreede) [reply python/python-adaguc-server/routeAdagucServer.py]: Removed the strict_slashes parameter

File: data/config/includes/DefaultStyles.include.xml (hunk 1 above)
Comment (ernstdevreede) [reply data/config/includes/DefaultStyles.include.xml:151]: I suggest to leave this as is and decide how to do this later, after we've had some more experience

**Comment by maartenplieger** [comment]
This one should be merged after the https://github.com/KNMI/adaguc-server/pull/248 ShadeInterval nearest PR, as that one has some conflicts with this one.

File: adagucserverEC/CImgRenderPolylines.cpp [hunk 11]
```diff
@@ -143,29 +193,56 @@ void CImgRenderPolylines::render(CImageWarper *imageWarper, CDataSource *dataSou
               cnt++;
             } else {
               CDBDebug("status: %d %d [%f,%f]", status, j, tprojectedX, tprojectedY);
```
Comment (maartenplieger) [inline adagucserverEC/CImgRenderPolylines.cpp]: Should we not remove all these Debug statements?

File: adagucserverEC/CImgRenderPolylines.cpp [hunk 12]
```diff
@@ -143,29 +193,56 @@ void CImgRenderPolylines::render(CImageWarper *imageWarper, CDataSource *dataSou
...
+              if (it != feature->getFp()->end()) {
+                featureId.print(it->second->toString(featureStyle.propertyFormat));
+              } else {
+                featureId = feature->getId();
+              }
+              CColor labelColor(featureStyle.fontColor);
+              drawImage->drawCenteredTextNoOverlap(dlon, height - dlat, featureStyle.fontFile.c_str(), featureStyle.fontSize, featureStyle.angle, featureStyle.padding, featureId, labelColor,
+                                                   noOverlap, rects);
+            } else {
+              CDBDebug("Status: %d centroid [%f,%f]", status, centroidX, centroidY);
```
Comment (maartenplieger) [inline adagucserverEC/CImgRenderPolylines.cpp]: Remove or make it depend on a macro setting

File: adagucserverEC/CImgRenderPolylines.cpp [hunk 13]
```diff
@@ -143,29 +193,56 @@ void CImgRenderPolylines::render(CImageWarper *imageWarper, CDataSource *dataSou
...
+            }
+            firstPolygon = false;
+          } else {
+            CDBDebug("2nd polygon for %d", cnt);
```
Comment (maartenplieger) [inline adagucserverEC/CImgRenderPolylines.cpp]: Remove or make it depend on a macro setting

File: adagucserverEC/CImgRenderPolylines.cpp [hunk 14]
```diff
@@ -143,29 +193,56 @@ void CImgRenderPolylines::render(CImageWarper *imageWarper, CDataSource *dataSou
...
-            //                       projectedX[j]=dlon;
-            //                       projectedY[j]=height-dlat;
           }
 
-          //                    CDBDebug("Draw polygon: %d points (%d)", cnt, numPoints);
+          // CDBDebug("Draw polygon: %d points (%d)", cnt, numPoints);
           drawImage->poly(projectedX, projectedY, cnt, drawPointLineWidth, drawPointLineColor2, drawPointLineColor2, true, false);
-          //                    break;
+          // Determine centroid for first polygon.
+          if (firstPolygon) {
```
Comment (maartenplieger) [inline adagucserverEC/CImgRenderPolylines.cpp:194]: Would it be possible to split this extra functionalities over different files?

Not really a big issue to put it in like this, just a question.

File: adagucserverEC/CImgRenderPolylines.cpp [hunk 15]
```diff
@@ -229,7 +300,12 @@ void CImgRenderPolylines::render(CImageWarper *imageWarper, CDataSource *dataSou
 #ifdef MEASURETIME
         StopWatch_Stop("Feature drawn %d", featureIndex);
 #endif
-        featureIndex++;
+      }
+      CDBDebug("Drawing %d rects", rects.size());
```
Comment (maartenplieger) [inline adagucserverEC/CImgRenderPolylines.cpp]: Remove or make it depend on a macro setting

File: doc/configuration/FeatureInterval.md [hunk 16]
```diff
@@ -3,12 +3,29 @@ FeatureInterval (match, matchid, label, bgcolor, fillcolor)
...
+-   labelfontfile - Optional, the TrueType fontfile to be used for drawing the label
+-   labelcolor - Optional, the color to be used for drawing the label
+-   labelproperytname- Optional, determines which GeoJSON property is used for the label
+-   labelpropertyformat - Optional, determines the printf format string used to convert the labelpropertyname into a string (default %s)
+-   labelangle - Optional, the angle at which the label will be drawn (in degrees), default 0 (=horizontal)
+-   labelpadding - Optional, the padding in pixels to be added around the bounding boxes of labels when not plotting overlapping labels
+
+The drawing of overlapping labels can be enabled or disabled with a [RenderSetting](./RenderSettings.md):
+-   featuresoverlap - Optional, values true or false, default false. Enable or disable overlapping labels, for readability
+-   randomizefeatures - Optional, values true or false, default false. Start drawing of list of labels at a random start point, so that not always the same overlapping labels are being skipped.
```
Comment (maartenplieger) [inline doc/configuration/FeatureInterval.md:28]: Could you link you your test configuration and test image or put in a screenshot with a small config template? That way you can immediately see how it works.

The configuration in your test and the image in expected outputs from your test are handy to re-use here.

File: adagucserverEC/CImgRenderPolylines.cpp (hunk 11 above)
Comment (ernstdevreede) [reply adagucserverEC/CImgRenderPolylines.cpp]: Done

File: adagucserverEC/CImgRenderPolylines.cpp (hunk 12 above)
Comment (ernstdevreede) [reply adagucserverEC/CImgRenderPolylines.cpp]: Done

File: adagucserverEC/CImgRenderPolylines.cpp (hunk 13 above)
Comment (ernstdevreede) [reply adagucserverEC/CImgRenderPolylines.cpp]: Done

File: adagucserverEC/CImgRenderPolylines.cpp (hunk 15 above)
Comment (ernstdevreede) [reply adagucserverEC/CImgRenderPolylines.cpp]: Done

File: adagucserverEC/CImgRenderPolylines.cpp (hunk 14 above)
Comment (ernstdevreede) [reply adagucserverEC/CImgRenderPolylines.cpp:194]: A bit too much work now, I feel.

File: doc/configuration/FeatureInterval.md (hunk 16 above)
Comment (ernstdevreede) [reply doc/configuration/FeatureInterval.md:28]: Done.

**Review by maartenplieger — APPROVED** [review]
OK
